### PR TITLE
Added filtering of allocations by several attributes

### LINKED ISSUE #30: Implement filtering in allocations API by attributes
Currently the invoicing tools make an API request to `/api/allocations?all=true`. This responds with ALL the allocations, which doesn't scale.

We need to allow filtering by a specific attribute so that we can search for the allocation that we need.

### PR DESCRIPTION
**Pull request by QuanMPhm**
Closes #30  Specifically, I have added filtering by project PI email, project title, resource name, resource type, and allocation id. This feature is implemented by creating a mapping between the query arguements and the django model query arguements. More details in my [comments](https://github.com/nerc-project/coldfront-plugin-api/compare/main...QuanMPhm:30/allocation_api?expand=1#diff-2be8627f55e66e9de9bf3b91fd9f2d48179167ccdeee11ca3e5801696fdd4b1bR13) to the allocations viewset.

### CODE REVIEW

**Comment by knikolla** [comment]
This looks good at first glance, and I will do a more in-depth review after the craziness of the workshop has passed.

One thing that this is missing and I'm more interested in seeing implemented is a filter on attributes added to an allocation, such as `Allocated Project ID`.

**Comment by QuanMPhm** [comment]
I've added filtering by allocation attributes and by any allocation model attributes. More explanation [here](https://github.com/nerc-project/coldfront-plugin-api/pull/36/files#diff-2be8627f55e66e9de9bf3b91fd9f2d48179167ccdeee11ca3e5801696fdd4b1bR14)

File: src/coldfront_plugin_api/urls.py [hunk 1]
```diff
@@ -10,16 +11,68 @@
...
 
+        allocation_attr_prefix = "attr_"
+        query_split_char = ","
+
+        for query, val in query_params.items():
+            if query == "all":
+                continue
+            if query.startswith(allocation_attr_prefix):
+                attribute_query = query[len(allocation_attr_prefix) :].replace("_", " ")
+                val_list = val.split(query_split_char)
```
Comment (knikolla) [inline src/coldfront_plugin_api/urls.py:58]: I think all of these functions are offered by `QueryDict`

Have you tried [getlist()](https://docs.djangoproject.com/en/5.0/ref/request-response/#django.http.QueryDict.getlist)?

File: src/coldfront_plugin_api/urls.py [hunk 2]
```diff
@@ -10,16 +11,68 @@
...
+                q_query = Q(allocationattribute__value=val_list[0])
+                for i in range(1, len(val_list)):
```
Comment (knikolla) [inline src/coldfront_plugin_api/urls.py:60]: Why are you using range?

```python
for value in value_list:
     ...
```

is much cleaner.

File: src/coldfront_plugin_api/urls.py [hunk 3]
```diff
@@ -10,16 +11,68 @@
...
+
+                queryset = queryset.filter(
+                    q_query,
+                    allocationattribute__allocation_attribute_type__name=attribute_query,
+                )
+
+            else:
+                val_list = val.split(query_split_char)
+                q_query = Q(**{query: val_list[0]})
+                for i in range(1, len(val_list)):
```
Comment (knikolla) [inline src/coldfront_plugin_api/urls.py:71]: Why are you using range?

```python
for value in value_list:
     ...
```

is much cleaner.

File: src/coldfront_plugin_api/urls.py [hunk 4]
```diff
@@ -10,16 +11,68 @@
...
 
-        if self.request.query_params.get("all") == "true":
+        if query_params.get("all") == "true":
             queryset = Allocation.objects.all()
 
+        allocation_attr_prefix = "attr_"
+        query_split_char = ","
+
+        for query, val in query_params.items():
+            if query == "all":
```
Comment (knikolla) [inline src/coldfront_plugin_api/urls.py:54]: Why is this here?

File: src/coldfront_plugin_api/tests/unit/test_allocations.py [hunk 5]
```diff
@@ -18,6 +19,17 @@ def setUp(self) -> None:
 
         self.resource = self.new_resource(name="Devstack", auth_url="http://localhost")
 
+    @staticmethod
+    def new_allocation_attribute(allocation, attribute, value):
```
Comment (knikolla) [inline src/coldfront_plugin_api/tests/unit/test_allocations.py:23]: Could [this](https://github.com/nerc-project/coldfront-plugin-cloud/blob/114fd83de5b818f05f10317b4f299d9083e2248e/src/coldfront_plugin_cloud/utils.py#L21) method have worked? Why did you write a new one?

**Review by knikolla — CHANGES_REQUESTED** [review]

File: src/coldfront_plugin_api/urls.py (hunk 4 above)
Comment (QuanMPhm) [reply src/coldfront_plugin_api/urls.py:54]: Because the query parameter `all` is already used here to check if we should filter against active or all allocations. 
I suppose I could use the `all` query in the for loop, but I thought that would look a bit disorganized/unclear

File: src/coldfront_plugin_api/urls.py [hunk 6]
```diff
@@ -10,16 +11,68 @@
...
+    is possible by using the query parameters directly as Django query filters.
+
+    For example, to filter by the allocation's project PI email: "/api/allocations?project__pi__email=test@bu.edu"
+
+    Filtering using this method requires knowledge of the Allocation model schema and Django's filters
+    Documentation of Coldfront Allocation model: https://coldfront.readthedocs.io/en/latest/apidocs/allocations/
+    Documentation for Django's query syntax: https://docs.djangoproject.com/en/5.0/topics/db/queries/#lookups-that-span-relationships
+
+    To filter by allocation attribute (AA) (i.e Quota Limits CPU),
+    name the query as the AA (case insensitive), prefix with "attr_", and replace all space with "_"
```
[the remark this answers is not in the document]
Comment (QuanMPhm) [reply src/coldfront_plugin_api/urls.py:30]: I guess I wanted to avoid spaces, since it "felt" like something to avoid in general in coding. I'll have it changed per your suggestion

File: src/coldfront_plugin_api/urls.py (hunk 1 above)
Comment (QuanMPhm) [reply src/coldfront_plugin_api/urls.py:58]: I not sure which `QueryDict` functions you think could be applied here. `getlist()` returns the value of a specified key, but I'm not sure how it would be applied here.

File: src/coldfront_plugin_api/tests/unit/test_allocations.py (hunk 5 above)
Comment (QuanMPhm) [reply src/coldfront_plugin_api/tests/unit/test_allocations.py:23]: I realized that the utility function in `coldfront_plugin_cloud` didn't return the allocation attribute, which I need to perform assertions. I could probably still use that function, but it might result in messier assertions.

File: src/coldfront_plugin_api/urls.py (hunk 1 above)
Comment (knikolla) [reply src/coldfront_plugin_api/urls.py:58]: According to the documentation for getlist:

> Returns a list of the data with the requested key

Which means it would handle the splitting for us.

Comment (QuanMPhm) [reply src/coldfront_plugin_api/urls.py:58]: `getlist()` is used in the case where the same query parameter is named multiple times in the query string, i.e `?a=1&a=2`, in which case `getlist('a')` would return the list `[1,2]`. 

In this if statement, I am checking for query parameters starting with a specific prefix (i.e `?attr_OpenStack GPU Quota=1,2`), and split the query string with using a specifc character (in this code, its `,`). I see that in both cases, the end goal is to get a list, indicating the values we want to use when filtering allocations by a specific attributes.

I guess instead of doing string splitting ourselves to get the query parameter, you could format the query url as `?attr_OpenStack GPU Quota=1&attr_OpenStack GPU Quota=2`)`, and use `getlist()` there, but I thought that looked too long.

Comment (knikolla) [reply src/coldfront_plugin_api/urls.py:58]: I see. That is still preferable than doing the splitting ourselves.

Comment (knikolla) [reply src/coldfront_plugin_api/urls.py:58]: And can you also add a few tests to verify what happens when you try to filter for a non existing attribute?

Comment (knikolla) [reply src/coldfront_plugin_api/urls.py:58]: No, a 404 means a specific requested object was not found. Filtering isn't a request for a specific object, and therefore either a 400 Bad Request or an empty list is fine.

File: src/coldfront_plugin_api/urls.py [hunk 7]
```diff
@@ -10,16 +12,72 @@
...
+        allocation_attr_prefix = "attr_"
+
+        for query, val in query_params.items():
+            if query == "all":
+                continue
+            if query.startswith(allocation_attr_prefix):
+                attribute_query = query[len(allocation_attr_prefix) :]
+                val_list = query_params.getlist(query)
+                q_query = Q(allocationattribute__value=val_list[0])
+                for val in val_list[1:]:
```
Comment (naved001) [inline src/coldfront_plugin_api/urls.py:62]: what's `val_list` supposed to look like? what's in `val_list[0]` and what's in the rest of it that we are iterating over? I am a little confused.

Comment (QuanMPhm) [reply src/coldfront_plugin_api/urls.py:62]: `val_list` is the list of values given for a query parameter.
I.e if the user queried this endpoint: `/api/allocations?a=1&a=2&b=3`, note how the query parameter `a` is used twice, so the `val_list` for `a` will contain `[1,2]`, and for `b` it will be `[3]`.

Comment (naved001) [reply src/coldfront_plugin_api/urls.py:65]: @QuanMPhm I think a better way to do this might be:

```python
q_query = Q()   # An empty Q query

for val in val_list:
    q_query = query | Q(allocationattribute__value=val)
```

I think this reads better. Definitely test this first because what I was just playing around in the interpreter.

File: src/coldfront_plugin_api/urls.py [hunk 8]
```diff
@@ -10,16 +12,75 @@
...
+    For example, to filter by the allocation's project PI email: "/api/allocations?project__pi__email=test@bu.edu"
+
+    Filtering using this method requires knowledge of the Allocation model schema and Django's filters
+    Documentation of Coldfront Allocation model: https://coldfront.readthedocs.io/en/latest/apidocs/allocations/
+    Documentation for Django's query syntax: https://docs.djangoproject.com/en/5.0/topics/db/queries/#lookups-that-span-relationships
+
+    To filter by allocation attribute (AA) (i.e Quota Limits CPU),
+    name the query as the AA (case insensitive), prefixed with "attr_"
+
+    For all filters, you can also filter with conditional OR,
```
Comment (naved001) [inline src/coldfront_plugin_api/urls.py:33]: Was the requirement only to filter with `OR`? Is there a reason we don't want to support filtering with `AND`

Comment (QuanMPhm) [reply src/coldfront_plugin_api/urls.py]: The query is written so that you can filter allocations using conditional AND on two or more attributes. I.e This means it's possible to filter for allocations with a CPU quota of 2 AND RAM quota of 4G. I have updated the docstring to explain this functionality.

Kristi only told me to filter on model and allocation attributes, and I just decided that the OR and AND operations were sufficient so far.

File: src/coldfront_plugin_api/urls.py (hunk 7 above)
Comment (QuanMPhm) [reply src/coldfront_plugin_api/urls.py]: You're right, initializing with an empty Q query is fine for Django.

File: src/coldfront_plugin_api/urls.py (hunk 8 above)
Comment (naved001) [reply src/coldfront_plugin_api/urls.py]: > you can filter allocations using conditional AND on two or more attributes. I.e This means it's possible to filter for allocations with a CPU quota of 2 AND RAM quota of 4G. 

Sorry I don't understand this, how would that work? Looking at the query that's being built we are only OR-ing the individual queries. How would the URI look like when we filter for allocation with a "CPU quota of 2 AND RAM quota of 4G" vs "a CPU quota of 2 OR RAM quota of 4G"?

Comment (QuanMPhm) [reply src/coldfront_plugin_api/urls.py]: Sorry. I needed to be even more specific... 
The filter code is written so that you could perform conditional OR for different *values* of the same attribute 
(CPU quota = 2 OR 4), and to perform conditional AND for different *attributes* (CPU quota = 2 && RAM quota = 8G).
It doesn't make sense to implement conditional AND for values of the same attribute (a quota can't be two values at the same time), and as for conditional OR for different attributes, I simply haven't implemented it because I haven't found it necessary.

An example URL using conditional OR (CPU Quota = 2 OR 4 OR 8):
`api/allocations?attr_CPU Quota=2&attr_CPU Quota=4&attr_CPU Quota=8`

Using conditional AND (CPU Quota = 2 AND RAM Quota )
`api/allocations?attr_CPU Quota=2&attr_RAM Quota=8G`

Using conditional AND with OR (CPU Quota = 2 OR 4, AND Ram Quota = 4G OR 16G)
`api/allocations?attr_CPU Quota=2&attr_CPU Quota=4&attr_RAM Quota=4G&attr_RAM Quota=16G`

**Review by knikolla — APPROVED** [review]

File: src/coldfront_plugin_api/urls.py [hunk 9]
```diff
@@ -10,16 +12,78 @@
...
+                    q_query = q_query | Q(allocationattribute__value=val)
+
+                queryset = queryset.filter(
+                    q_query,
+                    allocationattribute__allocation_attribute_type__name__iexact=attribute_query,
+                )
+
+            else:
+                val_list = query_params.getlist(query)
+                q_query = Q()
```
Comment (naved001) [inline src/coldfront_plugin_api/urls.py:78]: optional suggestion: these 2 could be moved out of the if/else as we do them both in either case.

**Review by naved001 — APPROVED** [review]
